### ISSUE TITLE
Option to not load testthat helpers in `load_all`

### ISSUE THREAD

**Issue by pitakakariki** [hadley/devtools#1202]
I use testthat helpers to set options that are appropriate for automated testing, but not for interactive development.

e.g. my testthat helpers turn off progress bars and set default bootstrap replicates to a low number.

It would be useful to have an option not to run these files when I use `load_all`.

**Comment by hadley** [comment]
Those are options to your package?

**Comment by pitakakariki** [comment]
Yes, options within my package.

**Comment by pitakakariki** [comment]
Current workaround, if anyone else is having this problem: set the options in `test_aaa.R` instead.

**Comment by kforner** [comment]
This causes me problems too, and because my tests are run in parallel I can't use the workaround as is.

I'd like know why load_all sources the test helpers ???
You could imagine doing things are expensive in those helpers, both in CPU and in memory, that are only useful when executing tests, so why would you want to have them executed at load time ?

So an option or a parameter would be welcome.
Thanks.

**Comment by jimhester** [comment]
I think the idea behind `load_all()` is to have an interactive environment that mimics the environment your tests are run in.

An alternative workaround is to keep the expensive things in your helper, but surround it in 

```r
if (!interactive() || identical(Sys.getenv("NOT_CRAN"), "true")) {
...
}
```
devtools only sets `NOT_CRAN=true` when running tests interactively, so this code will only be run by `devtools::test()` or during R CMD check, not when calling `load_all()`.

**Comment by kforner** [comment]
> I think the idea behind load_all() is to have an interactive environment that mimics the environment your >
 tests are run in.

As a long-time user, I would say that for me the idea is to be able to load "quickly" a package from source, as if it was installed. Period. Loading a package via library() DO NOT source the test helpers.

Why force this behaviour ? Why not add an option or make a special function for people that want to run tests interactively, like load_for_testing() ? or instead make a lower level load_all function: load_package_from_source() ? 

The UNIX philosophy is to try to make one thing right. Why can't I just  have a function that loads the package ?

**Comment by gaborcsardi** [comment]
`load_all` also loads the non-exported functions by default, so it is already different from `library()`.

But I agree that loading the test helpers by default is a bit too much. Usually I don't have any helpers, so not a serious issue, but it is sometimes annoying.

**Comment by kforner** [comment]
> load_all also loads the non-exported functions by default, so it is already different from library().

Yes but you have the **export_all** param to make it work like library (and I do use this option). 

I could totally live with the "by default behaviour", what bothers me is that I have no way but hacking the function to get the behaviour I need.

**Comment by hadley** [comment]
No, the point of `load_all()` is to make it easy to interactively develop the package. This is a big help for people starting out with package development.

**Comment by kforner** [comment]
ok. In this case, would it be possible to have the load_package_from_source()/library_from_src() function, basically load_all without the sourcing of test helpers ? 
Thanks.

**Comment by gaborcsardi** [comment]
I would second @kforner on this. It would be nice to have an option to avoid the test helpers, as one don't always needs them. At least I mostly don't. :)

**Comment by kforner** [comment]
The reason I insist is that it breaks our dev workflow (and that it was already painful to identify the cause). Maybe the design is flawed, but in any case it involves quite an effort to fix it. It's so much easier for me to downgrade devtools in the meantime, and to wait for an option not to run those helpers, or to patch load_all if the option will not happen.

**Comment by jimhester** [comment]
Another option is to use `devtools::dev_mode()` and `devtools::install(quick = TRUE)`, which should give you relatively quick library installs that don't interfere with your normally installed packages.

**Comment by gaborcsardi** [comment]
I suppose so, but these are not quite the same and somewhat less convenient. 

Is there really anything against an option that defaults to the current behaviour?

**Comment by krlmlr** [comment]
One packages I'm working on installs itself into a temporary library during testing. I'd like to do this in a helper, but that would slow down loading.

What if testthat had a way to reliably determine if it's currently testing or not? The helper code could check this and react accordingly.

**Comment by krlmlr** [comment]
I was able to work around the problem with a multicore future:

In the helper:

```
slow_result_future <- future::multicore(compute_and_return_result())
```

In the test:

```
result <- future::value(slow_result_future)
```

The future is evaluated in a separate process which doesn't block the main R process. On Windows, you need to use to future::multiprocess() instead.

**Comment by hadley** [comment]
If it's really causing pain, I'll take a PR to add an option

**Comment by kforner** [comment]
wOOt. an option (as in getOption()), or a function parameter ?

**Comment by pitakakariki** [comment]
https://github.com/hadley/devtools/pull/1169

**Comment by r2evans** [comment]
@jimhester's suggestion doesn't work for me, when I want to differentiate between calling `load_all()` (load no helpers) and `test()` (load helpers), since the `NO_CRAN` env var is not a player. My (hackish) adaptation (that appears to work for me):

```r
if (! interactive() || any(grepl("\\bload_all\\b", as.character(sys.calls()[1])))) {
  # ...
}
```

**Comment by hadley** [comment]
Moved to pkgload. PR at https://github.com/r-lib/pkgload/pull/41
